beacon-chain/core/helpers/committee.go

```diff
@@ -181,94 +181,69 @@ func VerifyBitfield(bitfield []byte, committeeSize int) (bool, error) {
-//            if validator_index in committee[0]
-//        ]
-//        if len(selected_committees) > 0:
-//            validators = selected_committees[0][0]
-//            shard = selected_committees[0][1]
-//            first_committee_at_slot = crosslink_committees[0][0]  # List[ValidatorIndex]
-//            is_proposer = first_committee_at_slot[slot % len(first_committee_at_slot)] == validator_index
-//
-//            assignment = (validators, shard, slot, is_proposer)
-//            return assignment
+//    for slot in range(epoch_start_slot, epoch_start_slot + SLOTS_PER_EPOCH)
+//        offset = committees_per_slot * (slot % SLOTS_PER_EPOCH)
+//        slot_start_shard = (get_epoch_start_shard(state, epoch) + offset) % SHARD_COUNT
+//        for i in range(committees_per_slot):
+//            shard = (slot_start_shard + i) % SHARD_COUNT
+//            committee = get_crosslink_committee(state, epoch, shard)
+//            if validator_index in committee:
+//                return committee, shard, slot
 func CommitteeAssignment(
 	state *pb.BeaconState,
-	slot uint64,
-	validatorIndex uint64,
-	registryChange bool) ([]uint64, uint64, uint64, bool, error) {
-	var selectedCommittees []*cache.CommitteeInfo
+	epoch uint64,
+	validatorIndex uint64) ([]uint64, uint64, uint64, bool, error) {
 
-	wantedEpoch := slot / params.BeaconConfig().SlotsPerEpoch
-	prevEpoch := PrevEpoch(state)
-	nextEpoch := NextEpoch(state)
-
-	if wantedEpoch < prevEpoch || wantedEpoch > nextEpoch {
+	if epoch > NextEpoch(state) {
 		return nil, 0, 0, false, fmt.Errorf(
-			"epoch %d out of bounds: %d <= epoch <= %d",
-			wantedEpoch,
-			prevEpoch,
-			nextEpoch,
-		)
+			"epoch %d can't be greater than next epoch %d",
+			epoch, NextEpoch(state))
 	}
 
-	var cachedCommittees *cache.CommitteesInSlot
-	var err error
-	startSlot := StartSlot(wantedEpoch)
+	committeesPerSlot := EpochCommitteeCount(state, epoch) / params.BeaconConfig().SlotsPerEpoch
+	epochStartShard, err := EpochStartShard(state, epoch)
+	if err != nil {
+		return nil, 0, 0, false, fmt.Errorf(
+			"could not get epoch start shard: %v", err)
+	}
+	startSlot := StartSlot(epoch)
 	for slot := startSlot; slot < startSlot+params.BeaconConfig().SlotsPerEpoch; slot++ {
-
-		cachedCommittees, err = committeeCache.CommitteesInfoBySlot(slot)
-		if err != nil {
-			return []uint64{}, 0, 0, false, err
-		}
-		if cachedCommittees == nil {
-			crosslinkCommittees := []*CrosslinkCommittee{}
-			cachedCommittees = ToCommitteeCache(slot, crosslinkCommittees)
-			if err := committeeCache.AddCommittees(cachedCommittees); err != nil {
-				return []uint64{}, 0, 0, false, err
+		offset := committeesPerSlot * (slot % params.BeaconConfig().SlotsPerEpoch)
+		slotStatShard := (epochStartShard + offset) % params.BeaconConfig().ShardCount
+		for i := uint64(0); i < committeesPerSlot; i++ {
+			shard := (slotStatShard + i) % params.BeaconConfig().ShardCount
+			committee, err := CrosslinkCommitteeAtEpoch(state, epoch, shard)
```

Comment:
we can add a function crosslinkCommitteeMap that return map and save the next for loop for searching

Reply:
We do cache shuffled validator indices here so I think we are ok without the map. We can run some benchmarks after this PR to see whether the map is needed. https://github.com/prysmaticlabs/prysm/pull/2682

Don't really wanna add additional logic to mutate helper function unless it's absolutely needed